fix(forms): type NG_VALUE_ACCESSOR injection token correctly

Airblader opened the pull request:
fixes #29351

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

`NG_VALUE_ACCESSOR` is typed `ControlValueAccessor`. This conflicts with it being used as `multi: true` when injecting it programmatically through `Injector#get` as the function is typed to return a single `ControlValueAccessor`, even though it should be an array of them.

Issue Number: #29351


## What is the new behavior?

The injection token is now typed `ControlValueAccessor[]`.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


## Other information

I have not added any tests for this as it's purely a type information. This cannot be tested in an assertion kind of test, the only possibility would be a nop `Injector#get` call and making sure the TS compiler doesn't complain. This isn't a functional test, though, and I doubt we generally test that types are correct, so it seems unnecessary.

Airblader commented:
Regarding the two failed CI jobs:

1. Yes, technically the public API changes, but see my description above for why I believe this is not an issue here.
2. The failing lifecycle hook tests seem unrelated to this change(?)

kara reviewed (dismissed):
LGTM in general. 

A few things need to happen before this can be merged:
- [ ] Run public_api guard task to update goldens (cannot be merged while this test is red)
            - `yarn bazel run //tools/public_api_guard:forms_api.accept`
- [ ] Update commit message to add more documentation about the change
- [ ] [Kara] Run through Google3 TAP to see how much of a breaking change this would be

kara commented:
[presubmit](http://test/OCL:247103405:BASE:247103421:1557268409263:2169f3c4)

Airblader commented:
@kara The command you listed fails for me with like below. Nonetheless, I think I should have updated everything accordingly.

```
$ bazel run //tools/public_api_guard:forms_api.accept
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=140
INFO: Reading rc options for 'run' from /home/ingo/repos/angular/.bazelrc:
  Inherited 'build' options: --strategy=AngularTemplateCompile=worker --strategy=TypeScriptCompile=standalone --symlink_prefix=dist/ --nowatchfs --incompatible_strict_action_env --experimental_ui --define=compile=legacy
ERROR: Unrecognized option: --incompatible_strict_action_env
error Command failed with exit code 2.
```

AndrewKushnir commented:
Hi @Airblader,

This change will most likely be breaking, so we can see if we can include it into one of the next major releases (v11). Could you please rebase this PR when you get a chance, so that I can run a new presubmit to estimate the effort that would be required to adapt g3 apps code to this change?

Thank you,
Andrew

Airblader committed:
fix(forms): type NG_VALUE_ACCESSOR injection token as array

NG_VALUE_ACCESSOR is a multi injection token, users can and
should expect more than one ControlValueAccessor to be
available (and this is how it is used in @angular/forms).

This is now reflected in the definition of the injection token
by typing it as an array of ControlValueAccessor. The motivating
reason is that using the programmatic Injector api will now
type Injector#get correspondingly.

fixes #29351

BREAKING CHANGES

NG_VALUE_ACCESSOR is now typed as a readonly array rather than
a mutable scalar. It is used as a multi injection token and as
such it should always be expected that more than one accessor
may be returned.

Airblader commented:
@AndrewKushnir I've rebased it.

AndrewKushnir commented:
Thanks for the rebase @Airblader 👍 I've started [g3 presubmit](http://test/OCL:310593046:BASE:310593097:1588961989382:4823cb31) to estimate the effort that'd be required to land this change.

AndrewKushnir commented:
Hi @Airblader, just wanted to let you know that g3 presubmits went well. Since this change is potentially breaking, we can include that into the next major release (in v11, since the scope of breaking changes in v10 was finalized). I'll keep you updated on when master branch is available for v11 changes. Thank you.

jelbourn commented on goldens/public-api/forms/forms.d.ts:
```diff
@@ -347,7 +347,7 @@ export declare const NG_ASYNC_VALIDATORS: InjectionToken<(Function | Validator)[
 export declare const NG_VALIDATORS: InjectionToken<(Function | Validator)[]>;
 
-export declare const NG_VALUE_ACCESSOR: InjectionToken<ControlValueAccessor>;
+export declare const NG_VALUE_ACCESSOR: InjectionToken<ControlValueAccessor[]>;
```
Something related happened to come up on another thread today- Should this actually be `ReadonlyArray<ControlValueAccessor>` in order to prevent injecting code from mutating the array ([context](https://github.com/angular/angular/issues/37481#issuecomment-644843479))?

AndrewKushnir commented:
Hi @Airblader,

We've released the 10.1.0 version last week and master branch is now available for breaking changes that should be included into the next major release (v11).

- Could you please address Jeremy's comment here: https://github.com/angular/angular/pull/29723#discussion_r442616654?
- Rebase the PR to make sure the CI is green is the most up-to-date version of master branch.
- Please add a "BREAKING CHANGES" section to the commit message, explaining what might break in apps and some recommendation on how to resolve that.
- [Andrew] Re-run tests in Google's codebase and adapt apps code to this change as needed.

Thank you.

Airblader commented:
I've rebased and updated according to the comment. Hoping GitHub will let us know that it's all green. Thanks!

AndrewKushnir commented:
@Airblader, thanks for updating the PR, it looks like a couple CI jobs are failing ([lint](https://app.circleci.com/pipelines/github/angular/angular/21034/workflows/73517edc-6e24-4078-b50a-708a28bf4058/jobs/802982) and [test](https://app.circleci.com/pipelines/github/angular/angular/21034/workflows/73517edc-6e24-4078-b50a-708a28bf4058/jobs/802985)). Could you please have a look at the failures and update the code? Thank you.

Airblader commented:
@AndrewKushnir Should be fixed now.

AndrewKushnir commented on goldens/public-api/forms/forms.d.ts:
```diff
@@ -349,7 +349,7 @@ export declare const NG_ASYNC_VALIDATORS: InjectionToken<(Function | Validator)[
 export declare const NG_VALIDATORS: InjectionToken<(Function | Validator)[]>;
 
-export declare const NG_VALUE_ACCESSOR: InjectionToken<ControlValueAccessor>;
+export declare const NG_VALUE_ACCESSOR: InjectionToken<readonly ControlValueAccessor[]>;
```
Hi @Airblader,

Sorry I didn't mention that yesterday: I think the implementation that you have would work, but we usually have the `ReadonlyArray` type in these cases, i.e.:

```suggestion
export declare const NG_VALUE_ACCESSOR: InjectionToken<ReadonlyArray<ControlValueAccessor>>;
```

Could you please update the PR to use `ReadonlyArray` instead?

Thank you.

Airblader commented:
I had that, but the tests failed because they said it has to be like this.

AndrewKushnir commented:
@Airblader could you please try to apply the changes to the `packages/forms/src/directives/control_value_accessor.ts` file and run the following command after that (it should update the `goldens/public-api/forms/forms.d.ts` file, the `.d.ts` files are generated by the script):

```
yarn public-api:update
```

If there are failing tests after these changes, please let us know, we'll look into that.

Thank you.

Airblader commented:
@AndrewKushnir I updated it and ran the yarn command, which did not change the public-api to the same but instead retained `readonly X[]`.

AndrewKushnir commented:
[Presubmit](http://test/OCL:310593046:BASE:331469406:1600054664453:3886d7a) + [Global Presubmit](https://test.corp.google.com/OCL:310593046:BASE:331469519:1600054738746:d57b409d).

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:332757754:BASE:332757759:1600648436965:2f4818fd) + [Global Presubmit #2](https://test.corp.google.com/OCL:332757754:BASE:332757939:1600648563587:63fd4902).

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

petebacondarwin reviewed (approved):
Reviewed-for: public-api

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

AndrewKushnir commented:
Quick update:
- tests in Google's codebase went well (including Global run)
- there were no cases in Google's codebase that required a cleanup due to this change (no targets were broken)
- there are **no migration planned** for this change due to low probability of breakages due to this change
- this PR has all the necessary approvals now and I'm adding it to the merge queue

@Airblader thanks for contributing to Angular! 👍
